types/wlr_keyboard.c

```diff
@@ -124,6 +126,22 @@ void wlr_keyboard_init(struct wlr_keyboard *kb,
 	kb->repeat_info.delay = 600;
 }
 
+void wlr_keyboard_release_keys(struct wlr_keyboard *kb) {
+	// This is a while loop rather than a for loop because releasing keys
+	// can make some keys pressed
+	while (kb->num_keycodes > 0) {
+		struct timespec now;
+		clock_gettime(CLOCK_MONOTONIC, &now);
+		struct wlr_event_keyboard_key event = {
+			.time_msec = (int64_t)now.tv_sec * 1000 + now.tv_nsec / 1000000,
+			.keycode = kb->keycodes[0],
+			.update_state = false,
+			.state = WL_KEYBOARD_KEY_STATE_RELEASED,
+		};
+		wlr_keyboard_notify_key(kb, &event);
+	}
+}
+
 void wlr_keyboard_destroy(struct wlr_keyboard *kb) {
 	if (kb == NULL) {
```

Comment:
Could `wlr_keyboard_release_keys` be called inside `wlr_keyboard_destroy` instead?
That way it could be private.

Reply:
That would be best indeed. I'm not a fan of having keyboard-specific stuff inside the common code.

Reply:
`wlr_keyboard_destroy` is called after the `destroy` signal on the `wlr_input_device` is emitted.
If the keys are released after that signal, the original issue isn't fixed anymore.